src/connectors/collectionService.ts

```diff
@@ -0,0 +1,228 @@
+      )
+      .innerJoin('article', 'article.id', 'article_id')
+      .where({ collectionId, state: ARTICLE_STATE.active })
+      .orderBy('order', reversed ? 'desc' : 'asc')
+      .modify((builder: Knex.QueryBuilder) => {
+        if (skip !== undefined && Number.isFinite(skip)) {
+          builder.offset(skip)
+        }
+        if (take !== undefined && Number.isFinite(take)) {
+          builder.limit(take)
+        }
+      })
+    const totalCount = records.length === 0 ? 0 : +records[0].totalCount
+    return [records, totalCount]
+  }
+
+  public deleteCollections = async (
+    ids: readonly string[],
+    authorId: string
+  ): Promise<boolean> => {
+    if (ids.length === 0) {
+      return false
+    }
+    const collections = await this.findByIds(ids)
+
+    for (const collection of collections) {
+      if (!collection) {
+        throw new Error('Collection not found')
+      }
+      if (collection instanceof Error) {
+        throw new Error('Load collection error')
+      }
+      if (collection.authorId !== authorId) {
+        throw new Error('Author id not match')
+      }
+    }
+
+    await this.knex('collection_article').whereIn('collection_id', ids).del()
+    const result = await this.knex('collection')
+      .whereIn('id', ids)
+      .andWhere('author_id', authorId)
+      .del()
+    return result > 0
+  }
+
+  public deleteCollectionArticles = async (
+    collectionId: string,
+    articleIds: readonly string[]
+  ) =>
+    await this.knex('collection_article')
+      .where('collection_id', collectionId)
+      .whereIn('article_id', articleIds)
+      .del()
+
+  public findById = async (id: string) => await this.dataloader.load(id)
+
+  public findByIds = async (ids: readonly string[]) =>
+    await this.dataloader.loadMany(ids)
+
+  public addArticles = async (
+    collectionId: string,
+    articleIds: readonly string[]
+  ) => {
+    const res = await this.knex('collection_article')
+      .where('collection_id', collectionId)
+      .max('order')
+    const initOrder = res[0].max ? parseFloat(res[0].max) + 1 : 1
+    await this.knex('collection_article').insert(
+      articleIds.map((articleId, index) => ({
+        articleId,
+        collectionId,
+        order: initOrder + index,
+      }))
+    )
+  }
+
+  public reorderArticles = async (
+    collectionId: string,
+    moves: Array<{ articleId: string; newPosition: number }>
+  ) => {
+    const [collectionArticles, count] =
+      await this.findAndCountArticlesInCollection(collectionId, {})
+
+    if (
+      moves.some(({ newPosition }) => newPosition < 0 || newPosition >= count)
+    ) {
+      throw new Error('Invalid newPosition')
+    }
+
+    const articleIds = collectionArticles.map(({ articleId }) => articleId)
+    if (moves.some(({ articleId }) => !articleIds.includes(articleId))) {
+      throw new Error('Invalid articleId')
+    }
+
+    for (const { articleId, newPosition } of moves) {
```

Comment:
@pitb2022 can reuse this reordering algorithm for `optimisticResponse` on client-side.